<a href="https://colab.research.google.com/github/rezhaauliar/Tugas-Optimasi-Rezha/blob/main/multiclass_classification_colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Copyright 2020 Google LLC.

In [112]:
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Multiclass Classification

We previously created a binary classification model that determined if a piece of fruit was an orange or a grapefruit. There are many problems where binary classification can provide impactful solutions: spam or not spam in an email classifier, hit or hold in a blackjack simulation, buy or not in a stock market analysis. The list is basically endless.

There are other cases, however, where we want to make a decision across three or more classes. This is multiclass classification.

For many applications, multiclass classification can be broken down into many binary classification problems. These models employ a one-vs-all or one-vs-one strategy to create many binary classification tasks that are then aggregated into a multiclass classification model. Neural networks, decision trees, and k-nearest neighbors models are all capable of performing multiclass classification directly.

## The Dataset

For this unit we are going to use a classic machine learning dataset, the [Iris flower dataset](https://en.wikipedia.org/wiki/Iris_flower_data_set). This is a dataset that was used in 1936 by British biologist and statistician Ronald Fisher to classify iris flowers into one of three species based on four measurements:

- The length of the petals
- The width of the petals
- The length of the sepals (the green petal-looking bits that are found at the base of the petals)
- The width of the sepals

Conveniently, the iris dataset is built into the scikit-learn library, so it is readily available to us. Let's take a look:

In [113]:
from sklearn import datasets

iris_bunch = datasets.load_iris()
iris_bunch.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [114]:
iris_bunch['target']

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [115]:
iris_bunch['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

Scikit-learn datasets are usually delivered in the form of a dictionary-like object called a `Bunch`. This `Bunch` contains the following fields:

- *DESCR*: A string describing the dataset.
- *data*: An array containing the features we are using for classifying. In this case, it's the four measurements listed above for each of 150 plants.
- *feature_names*: Labels for the data.
- *filename*: the file that this data came from.
- *target*: the values that we are trying to classify these flowers into. In this case, since we are dealing with three species of iris, we use three numbers (0, 1 and 2) to identify each species.
- *target_names*: labels for the target values. In this case, 0 refers to the setosa species, 1 to the versicolor species, and 2 to the virginica species.

We'll create a list of columns that we'll use for our model.

In [116]:
FEATURES = iris_bunch['feature_names']
TARGET = 'species'

FEATURES, TARGET

(['sepal length (cm)',
  'sepal width (cm)',
  'petal length (cm)',
  'petal width (cm)'],
 'species')

Next we will load the feature and target data into a Pandas dataframe.

In [117]:
import pandas as pd

iris_df = pd.DataFrame(iris_bunch['data'], columns=FEATURES)
iris_df[TARGET] = iris_bunch['target']

iris_df.sample(10)

sepal length (cm)  sepal width (cm)  ...  petal width (cm)  species
24                 4.8               3.4  ...               0.2        0
63                 6.1               2.9  ...               1.4        1
56                 6.3               3.3  ...               1.6        1
18                 5.7               3.8  ...               0.3        0
135                7.7               3.0  ...               2.3        2
15                 5.7               4.4  ...               0.4        0
126                6.2               2.8  ...               1.8        2
4                  5.0               3.6  ...               0.2        0
1                  4.9               3.0  ...               0.2        0
65                 6.7               3.1  ...               1.4        1

[10 rows x 5 columns]

Let's take a look at a description of the data.

In [118]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  ...  petal width (cm)     species
count         150.000000        150.000000  ...        150.000000  150.000000
mean            5.843333          3.057333  ...          1.199333    1.000000
std             0.828066          0.435866  ...          0.762238    0.819232
min             4.300000          2.000000  ...          0.100000    0.000000
25%             5.100000          2.800000  ...          0.300000    0.000000
50%             5.800000          3.000000  ...          1.300000    1.000000
75%             6.400000          3.300000  ...          1.800000    2.000000
max             7.900000          4.400000  ...          2.500000    2.000000

[8 rows x 5 columns]

There are 150 data points. No columns seem to be missing data and no values seem to be too far out of expected ranges. For example, there are no zero or negative lengths or widths, and the length and width values fall well within what we'd expect for a tulip.

We are interested in using the measurement features to predict the species of an iris. Let's take a closer look at the values we'll be predicting.

In this case we'll group by our 'species' feature and get a count of each species in our dataset.

In [119]:
iris_df.groupby(TARGET).agg('count')

sepal length (cm)  ...  petal width (cm)
species                     ...                  
0                       50  ...                50
1                       50  ...                50
2                       50  ...                50

[3 rows x 4 columns]

We have 50 examples of each species of iris, and overall we have only 150 samples. This presents two challenges. First, we don't have much data to actually build a model from. Second, the data that we do have is evenly distributed over class types. We might want to make sure that we train over the same distribution.

Luckily, there are solutions to both of these issues!

When we have data that has some weighted distribution across classes, we can do a **stratified split** to ensure that every class appears proportionally in our training data.

When we don't have enough data to properly train a model and don't feel that we can pull training data away, we can do a **k-fold cross validation** in order to utilize all of our data for training, while still trying to minimize model overfitting.

## Stratified Split

Let's first split off a set of data to use for our final model testing. We can use scikit-learn's `train_test_split` function to do this.

Since we have so little data, we'll only hold out 10% of the data for the final test.

After we make the split, we can see how many data points we will train off of for each class.

In [120]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    iris_df[FEATURES],
    iris_df[TARGET],
    test_size=0.1,
    random_state=45)

y_train.groupby(y_train).count()

species
0    43
1    50
2    42
Name: species, dtype: int64

Yikes! We kept 50 data points for training class 1. That means we left none for final testing:

In [121]:
y_test.groupby(y_test).count()

species
0    7
2    8
Name: species, dtype: int64

### Exercise 1: Stratified Train Test Split

We risk not holding out a data point for every class if we don't stratify our train test split. Rewrite the split above to create a stratified split. (If you don't remember how, try looking at the [documentation for `train_test_split`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) and finding the argument that can be used to stratify the data.) When you are done, there should be 45 data points for each class in the training data and five data points for each class in the testing data. Print the counts to verify.

#### **Student Solution**

In [122]:
# Your Code Goes Here
X_train, X_test, y_train, y_test = train_test_split(
    iris_df[FEATURES],
    iris_df[TARGET],
    test_size=0.2,
    stratify=iris_df[TARGET])

print(y_train.groupby(y_train).count())
print(y_test.groupby(y_test).count())

species
0    40
1    40
2    40
Name: species, dtype: int64
species
0    10
1    10
2    10
Name: species, dtype: int64


---

## Cross-Validation

Another problem we have is that we have very little data to work with. We only had 150 data points in total and are only going to train using 135 of those data points. If we are going to be hyperparameter tuning, we'll need a test and validation holdout, which will leave us very little data to train on.

One way to get around this is to use cross-validation. Cross-validation splits the data into a fixed number of tranches and trains on `n-1` of the tranches. Then it calculates a score using the holdout tranche. It does this repeatedly, holding out one tranche of data for each training pass. By looking at the mean of the scores for each training pass, you can get an idea of how well your model performs without having to specify a test dataset.

The `cross_val_score` method is used to perform the cross-validation. In the example below, we divide the data into five tranches and get five scores.

Since we are cross-validating with a classifier, scikit-learn automatically performs stratified splits for us.

In [123]:
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import cross_val_score

estimator = SGDClassifier()

scores = cross_val_score(
    estimator,
    X_train,
    y_train,
    cv=5
)

scores

array([0.875     , 0.66666667, 0.95833333, 0.91666667, 0.75      ])

We can now find the mean score.

In [124]:
scores.mean()

0.8333333333333333

What does this score represent, though? It turns out that it uses the default scoring method for the classifier that we used. In this case we used the [`SGDClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html), which reports accuracy by default.

Also note that the estimator isn't trained after running cross-validation. You can run cross-validation to test different data preprocessing pipelines and hyperparameters. Once you are happy with a specific setup, you'll need to train the model with the chosen pipeline and parameters.

### Exercise 2: F1 Scoring

What if we wanted to use F1 for our scoring metric instead of accuracy?

Run `cross_val_score` on an `SGDClassifier` and get the F1 score. Check out the documentation for [`cross_val_score`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html), [`make_scorer`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html), and [`f1_score`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html) for clues.

#### **Student Solution**

In [125]:
# Your Code Goes here
import pandas as pd

from sklearn import datasets
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

iris_bunch = datasets.load_iris()
iris_bunch.keys()

FEATURES = iris_bunch['feature_names']
TARGET = 'species'

iris_df = pd.DataFrame(iris_bunch['data'], columns=FEATURES)
iris_df[TARGET] = iris_bunch['target']

X_train, X_test, y_train, y_test = train_test_split(
    iris_df[FEATURES],
    iris_df[TARGET], 
    test_size=0.1,
    stratify=iris_df[TARGET])

# Solution Starts Here

estimator = SGDClassifier()

scores = cross_val_score(
    estimator,
    X_train[FEATURES],
    y_train,
    cv=5,
    scoring=make_scorer(f1_score, average='micro')
)


scores.mean()

0.7777777777777779

---

# The Model Pipeline

Since we are now using cross-validation to train the model, we can use our testing holdout data as a final validation. Let's make that clear by renaming the data.

In [126]:
X_validation = X_test
y_validation = y_test

Now we can work on tuning the model and the model pipeline.

Let's first look back at the data going into the model:



In [127]:
iris_df[FEATURES].describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

The data is all in the same order of magnitude, but columns like 'sepal length (cm)' are considerably larger than columns like 'petal width (cm)'.

We need to perform some preprocessing to get the data into a more uniform shape before feeding it to the model. To do that we'll use the [`StandardScaler`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html), which removes the mean and subtracts the unit variance from each column of data.

To use the `StandardScaler`, we create the object, `fit()` the data, and then `transform()`.

In [128]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(iris_df[FEATURES])

pd.DataFrame(
    scaler.transform(iris_df[FEATURES]),
    columns=FEATURES
).describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count       1.500000e+02      1.500000e+02       1.500000e+02      1.500000e+02
mean       -1.690315e-15     -1.842970e-15      -1.698641e-15     -1.409243e-15
std         1.003350e+00      1.003350e+00       1.003350e+00      1.003350e+00
min        -1.870024e+00     -2.433947e+00      -1.567576e+00     -1.447076e+00
25%        -9.006812e-01     -5.923730e-01      -1.226552e+00     -1.183812e+00
50%        -5.250608e-02     -1.319795e-01       3.364776e-01      1.325097e-01
75%         6.745011e-01      5.586108e-01       7.627583e-01      7.906707e-01
max         2.492019e+00      3.090775e+00       1.785832e+00      1.712096e+00

You can see in the output of `describe()` that the data now all has a standard deviation that approaches one.

We need to perform this preprocessing to features before training the model and before getting predictions. It can be error-prone to try to remember to do this. To make the task easier, we can create an estimator [`Pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) that applies our transformations and calls our estimator.

In [129]:
from sklearn.pipeline import Pipeline

estimator = Pipeline(
  steps=[
    ['scale', StandardScaler()],
    ['classifier', SGDClassifier()],
  ]
)

scores = cross_val_score(
    estimator,
    X_train[FEATURES],
    y_train,
    cv=5,
)

scores.mean()

0.9407407407407409

Scaling gave us a considerable jump in accuracy score. Hopefully you see similar results.

### Exercise 3: Final Validation

Our accuracy results were pretty good, so we aren't going to do any more hyperparameter tuning in this lab. Before we declare victory, though, we should find the F1 score of our validation data. Using our estimator pipeline, calculate the F1 score for `X_validation`.

In [130]:
# Your Code Goes Here
import pandas as pd

from sklearn import datasets
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

iris_bunch = datasets.load_iris()
iris_bunch.keys()

FEATURES = iris_bunch['feature_names']
TARGET = 'species'

iris_df = pd.DataFrame(iris_bunch['data'], columns=FEATURES)
iris_df[TARGET] = iris_bunch['target']

X_train, X_validation, y_train, y_validation = train_test_split(
    iris_df[FEATURES],
    iris_df[TARGET], 
    test_size=0.1,
    stratify=iris_df[TARGET])

estimator = Pipeline(
  steps=[
    ['scale', StandardScaler()],
    ['classifier', SGDClassifier()],
  ]
)

scores = cross_val_score(
    estimator,
    X_train[FEATURES],
    y_train,
    cv=5,
    # We calculated F1 here too. This isn't required.
    scoring=make_scorer(f1_score, average='micro')
)

# Solution Starts Here
estimator.fit(X_train, y_train)
predictions = estimator.predict(X_validation)
f1_score(y_validation, predictions, average='micro')

1.0

In [131]:
#importing confusion matrix
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_validation, predictions)
print('Confusion Matrix\n')
print(confusion)

#importing accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print('\nAccuracy: {:.2f}\n'.format(accuracy_score(y_validation, predictions)))

print('Micro Precision: {:.2f}'.format(precision_score(y_validation, predictions, average='micro')))
print('Micro Recall: {:.2f}'.format(recall_score(y_validation, predictions, average='micro')))
print('Micro F1-score: {:.2f}\n'.format(f1_score(y_validation, predictions, average='micro')))

print('Macro Precision: {:.2f}'.format(precision_score(y_validation, predictions, average='macro')))
print('Macro Recall: {:.2f}'.format(recall_score(y_validation, predictions, average='macro')))
print('Macro F1-score: {:.2f}\n'.format(f1_score(y_validation, predictions, average='macro')))

print('Weighted Precision: {:.2f}'.format(precision_score(y_validation, predictions, average='weighted')))
print('Weighted Recall: {:.2f}'.format(recall_score(y_validation, predictions, average='weighted')))
print('Weighted F1-score: {:.2f}'.format(f1_score(y_validation, predictions, average='weighted')))

from sklearn.metrics import classification_report
print('\nClassification Report\n')
print(classification_report(y_validation, predictions, target_names=['Class 1', 'Class 2', 'Class 3']))

Confusion Matrix

[[5 0 0]
 [0 5 0]
 [0 0 5]]

Accuracy: 1.00

Micro Precision: 1.00
Micro Recall: 1.00
Micro F1-score: 1.00

Macro Precision: 1.00
Macro Recall: 1.00
Macro F1-score: 1.00

Weighted Precision: 1.00
Weighted Recall: 1.00
Weighted F1-score: 1.00

Classification Report

              precision    recall  f1-score   support

     Class 1       1.00      1.00      1.00         5
     Class 2       1.00      1.00      1.00         5
     Class 3       1.00      1.00      1.00         5

    accuracy                           1.00        15
   macro avg       1.00      1.00      1.00        15
weighted avg       1.00      1.00      1.00        15



---

# Exercise 4: Winemaker Identification

Scikit-learn comes prepackaged with many toy datasets. These can be found in the [`sklearn.datasets` package](https://scikit-learn.org/stable/datasets/index.html). In this exercise we'll be working with the [wine dataset](https://scikit-learn.org/stable/datasets/index.html#wine-dataset).

The dataset contains information about the properties of wines produced by three different producers. The grapes that the producers used all come from the same region.

The columns are:

* alcohol
* malic_acid
* ash
* alcalinity_of_ash
* magnesium
* total_phenols
* flavanoids
* nonflavanoid_phenols
* proanthocyanins
* color_intensity
* hue
* od280/od315_of_diluted_wines
* proline

The target column is a 0, 1, or 2. Each number represents a different producer.

Your task in this exercise is to create a classifier that can identify the producer based on the wine properties.

Use as many code blocks as necessary to examine the data and build and validate your model. Document your process using text blocks and/or comments in your code.

**Student Solution**

In [132]:
# Your Code Goes Here
import pandas as pd

from sklearn import datasets
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

wine_bunch = datasets.load_wine()
wine_bunch.keys()


dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

In [133]:

FEATURES = wine_bunch['feature_names']
TARGET = 'species'

wine_df = pd.DataFrame(wine_bunch['data'], columns=FEATURES)
wine_df[TARGET] = wine_bunch['target']


In [134]:
wine_df.head(10)

alcohol  malic_acid   ash  ...  od280/od315_of_diluted_wines  proline  species
0    14.23        1.71  2.43  ...                          3.92   1065.0        0
1    13.20        1.78  2.14  ...                          3.40   1050.0        0
2    13.16        2.36  2.67  ...                          3.17   1185.0        0
3    14.37        1.95  2.50  ...                          3.45   1480.0        0
4    13.24        2.59  2.87  ...                          2.93    735.0        0
5    14.20        1.76  2.45  ...                          2.85   1450.0        0
6    14.39        1.87  2.45  ...                          3.58   1290.0        0
7    14.06        2.15  2.61  ...                          3.58   1295.0        0
8    14.83        1.64  2.17  ...                          2.85   1045.0        0
9    13.86        1.35  2.27  ...                          3.55   1045.0        0

[10 rows x 14 columns]

In [135]:
wine_df.describe()

alcohol  malic_acid  ...      proline     species
count  178.000000  178.000000  ...   178.000000  178.000000
mean    13.000618    2.336348  ...   746.893258    0.938202
std      0.811827    1.117146  ...   314.907474    0.775035
min     11.030000    0.740000  ...   278.000000    0.000000
25%     12.362500    1.602500  ...   500.500000    0.000000
50%     13.050000    1.865000  ...   673.500000    1.000000
75%     13.677500    3.082500  ...   985.000000    2.000000
max     14.830000    5.800000  ...  1680.000000    2.000000

[8 rows x 14 columns]

---

In [136]:
wine_df.groupby(TARGET).agg('count')

alcohol  malic_acid  ash  ...  hue  od280/od315_of_diluted_wines  proline
species                            ...                                            
0             59          59   59  ...   59                            59       59
1             71          71   71  ...   71                            71       71
2             48          48   48  ...   48                            48       48

[3 rows x 13 columns]

In [137]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    wine_df[FEATURES],
    wine_df[TARGET],
    test_size=0.2,
    random_state=50)

y_train.groupby(y_train).count()

species
0    46
1    57
2    39
Name: species, dtype: int64

In [138]:
y_test.groupby(y_test).count()

species
0    13
1    14
2     9
Name: species, dtype: int64

In [139]:
# Your Code Goes Here
X_train, X_test, y_train, y_test = train_test_split(
    wine_df[FEATURES],
    wine_df[TARGET],
    test_size=0.3,
    stratify=wine_df[TARGET])

print(y_train.groupby(y_train).count())
print(y_test.groupby(y_test).count())

species
0    41
1    50
2    33
Name: species, dtype: int64
species
0    18
1    21
2    15
Name: species, dtype: int64


In [140]:
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import cross_val_score

estimator = SGDClassifier()

scores = cross_val_score(
    estimator,
    X_train,
    y_train,
    cv=5
)

scores

array([0.6       , 0.6       , 0.56      , 0.36      , 0.66666667])

In [141]:
scores.mean()

0.5573333333333333

In [142]:
# Your Code Goes here
import pandas as pd

from sklearn import datasets
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

wine_bunch = datasets.load_iris()
wine_bunch.keys()

FEATURES = wine_bunch['feature_names']
TARGET = 'species'

wine_df = pd.DataFrame(wine_bunch['data'], columns=FEATURES)
wine_df[TARGET] = wine_bunch['target']

X_train, X_test, y_train, y_test = train_test_split(
    wine_df[FEATURES],
    wine_df[TARGET], 
    test_size=0.1,
    stratify=wine_df[TARGET])

# Solution Starts Here

estimator = SGDClassifier()

scores = cross_val_score(
    estimator,
    X_train[FEATURES],
    y_train,
    cv=5,
    scoring=make_scorer(f1_score, average='micro')
)


scores.mean()

0.711111111111111

In [143]:
X_validation = X_test
y_validation = y_test

In [144]:
iris_df[FEATURES].describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

In [145]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(wine_df[FEATURES])

pd.DataFrame(
    scaler.transform(wine_df[FEATURES]),
    columns=FEATURES
).describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count       1.500000e+02      1.500000e+02       1.500000e+02      1.500000e+02
mean       -1.690315e-15     -1.842970e-15      -1.698641e-15     -1.409243e-15
std         1.003350e+00      1.003350e+00       1.003350e+00      1.003350e+00
min        -1.870024e+00     -2.433947e+00      -1.567576e+00     -1.447076e+00
25%        -9.006812e-01     -5.923730e-01      -1.226552e+00     -1.183812e+00
50%        -5.250608e-02     -1.319795e-01       3.364776e-01      1.325097e-01
75%         6.745011e-01      5.586108e-01       7.627583e-01      7.906707e-01
max         2.492019e+00      3.090775e+00       1.785832e+00      1.712096e+00

In [146]:
from sklearn.pipeline import Pipeline

estimator = Pipeline(
  steps=[
    ['scale', StandardScaler()],
    ['classifier', SGDClassifier()],
  ]
)

scores = cross_val_score(
    estimator,
    X_train[FEATURES],
    y_train,
    cv=5,
)

scores.mean()

0.9259259259259259

In [147]:
# Your Code Goes Here
import pandas as pd

from sklearn import datasets
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

wine_bunch = datasets.load_iris()
wine_bunch.keys()

FEATURES = wine_bunch['feature_names']
TARGET = 'species'

wine_df = pd.DataFrame(wine_bunch['data'], columns=FEATURES)
wine_df[TARGET] = wine_bunch['target']

X_train, X_validation, y_train, y_validation = train_test_split(
    wine_df[FEATURES],
    wine_df[TARGET], 
    test_size=0.2,
    stratify=wine_df[TARGET])

estimator = Pipeline(
  steps=[
    ['scale', StandardScaler()],
    ['classifier', SGDClassifier()],
  ]
)

scores = cross_val_score(
    estimator,
    X_train[FEATURES],
    y_train,
    cv=5,
    # We calculated F1 here too. This isn't required.
    scoring=make_scorer(f1_score, average='micro')
)

# Solution Starts Here
estimator.fit(X_train, y_train)
predictions = estimator.predict(X_validation)
f1_score(y_validation, predictions, average='micro')

1.0

In [148]:
#importing confusion matrix
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_validation, predictions)
print('Confusion Matrix\n')
print(confusion)

#importing accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print('\nAccuracy: {:.2f}\n'.format(accuracy_score(y_validation, predictions)))

print('Micro Precision: {:.2f}'.format(precision_score(y_validation, predictions, average='micro')))
print('Micro Recall: {:.2f}'.format(recall_score(y_validation, predictions, average='micro')))
print('Micro F1-score: {:.2f}\n'.format(f1_score(y_validation, predictions, average='micro')))

print('Macro Precision: {:.2f}'.format(precision_score(y_validation, predictions, average='macro')))
print('Macro Recall: {:.2f}'.format(recall_score(y_validation, predictions, average='macro')))
print('Macro F1-score: {:.2f}\n'.format(f1_score(y_validation, predictions, average='macro')))

print('Weighted Precision: {:.2f}'.format(precision_score(y_validation, predictions, average='weighted')))
print('Weighted Recall: {:.2f}'.format(recall_score(y_validation, predictions, average='weighted')))
print('Weighted F1-score: {:.2f}'.format(f1_score(y_validation, predictions, average='weighted')))

from sklearn.metrics import classification_report
print('\nClassification Report\n')
print(classification_report(y_validation, predictions, target_names=['Class 1', 'Class 2', 'Class 3']))

Confusion Matrix

[[10  0  0]
 [ 0 10  0]
 [ 0  0 10]]

Accuracy: 1.00

Micro Precision: 1.00
Micro Recall: 1.00
Micro F1-score: 1.00

Macro Precision: 1.00
Macro Recall: 1.00
Macro F1-score: 1.00

Weighted Precision: 1.00
Weighted Recall: 1.00
Weighted F1-score: 1.00

Classification Report

              precision    recall  f1-score   support

     Class 1       1.00      1.00      1.00        10
     Class 2       1.00      1.00      1.00        10
     Class 3       1.00      1.00      1.00        10

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30

